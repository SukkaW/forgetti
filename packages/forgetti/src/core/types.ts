import type * as t from '@babel/types';
import type * as babel from '@babel/core';
import type {
  HookRegistration,
  ImportRegistration,
  Options,
  Preset,
} from './presets';

export type ComponentNode =
  | t.ArrowFunctionExpression
  | t.FunctionExpression
  | t.FunctionDeclaration;

export interface State extends babel.PluginPass {
  opts: Options;
}

export interface StateContext {
  imports: Map<string, t.Identifier>;
  registrations: {
    named: {
      hooks: Map<t.Identifier, HookRegistration>;
      hocs: Map<t.Identifier, ImportRegistration>;
    };
    namespace: {
      hooks: Map<t.Identifier, HookRegistration[]>;
      hocs: Map<t.Identifier, ImportRegistration[]>;
    };
  };
  preset: Preset;
  filters: {
    component: RegExp;
    hook?: RegExp;
  };
}

export interface OptimizedExpression {
  expr: t.Expression;
  deps?: t.Expression | t.Expression[];
  constant?: boolean;
}
