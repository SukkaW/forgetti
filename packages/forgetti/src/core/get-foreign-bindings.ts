import type * as babel from '@babel/core';
import * as t from '@babel/types';

export function isForeignBinding(
  source: babel.NodePath,
  current: babel.NodePath,
  name: string,
): boolean {
  if (current.scope.hasGlobal(name)) {
    return false;
  }
  if (source === current) {
    return true;
  }
  if (current.scope.hasOwnBinding(name)) {
    return false;
  }
  if (current.parentPath) {
    return isForeignBinding(source, current.parentPath, name);
  }
  return true;
}

function isInTypescript(path: babel.NodePath): boolean {
  let parent = path.parentPath;
  while (parent) {
    if (t.isTypeScript(parent.node) && !t.isExpression(parent.node)) {
      return true;
    }
    parent = parent.parentPath;
  }
  return false;
}

export default function getForeignBindings(path: babel.NodePath): t.Identifier[] {
  const identifiers = new Set<string>();
  path.traverse({
    Expression(p) {
      // Check identifiers that aren't in a TS expression
      if (
        p.node.type === 'Identifier'
        && !isInTypescript(p)
        && isForeignBinding(path, p, p.node.name)
      ) {
        identifiers.add(p.node.name);
      }
      if (p.node.type === 'JSXElement') {
        switch (p.node.openingElement.name.type) {
          case 'JSXIdentifier': {
            const literal = p.node.openingElement.name.name;
            if (/^[A-Z_]/.test(literal) && isForeignBinding(path, p, literal)) {
              identifiers.add(literal);
            }
          }
            break;
          case 'JSXMemberExpression': {
            let base: t.JSXMemberExpression | t.JSXIdentifier = p.node.openingElement.name;
            while (base.type === 'JSXMemberExpression') {
              base = base.object;
            }
            if (isForeignBinding(path, p, base.name)) {
              identifiers.add(base.name);
            }
          }
            break;
          default:
            break;
        }
      }
    },
  });

  const result: t.Identifier[] = [];
  for (const identifier of identifiers) {
    result.push(t.identifier(identifier));
  }
  return result;
}
