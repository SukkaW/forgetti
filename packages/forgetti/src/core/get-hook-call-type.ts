/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type * as babel from '@babel/core';
import * as t from '@babel/types';
import { isHookName, isPathValid } from './checks';
import type { HookIdentity } from './presets';
import unwrapNode from './unwrap-node';
import type { StateContext } from './types';

type HookCallType =
  | HookIdentity
  | 'custom'
  | 'none';

export function getHookCallType(
  ctx: StateContext,
  path: babel.NodePath<t.CallExpression | t.OptionalCallExpression>,
): HookCallType {
  const callee = path.get('callee');
  if (isPathValid(callee, t.isV8IntrinsicIdentifier)) {
    return 'none';
  }
  const trueID = unwrapNode(callee.node, t.isIdentifier);
  if (trueID) {
    const binding = path.scope.getBindingIdentifier(trueID.name);
    if (binding) {
      const registration = ctx.registrations.named.hooks.get(binding);
      if (registration) {
        return registration.type;
      }
    }
    return isHookName(ctx, trueID) ? 'custom' : 'none';
  }
  // Check if callee is potentially a namespace import
  const trueMember = unwrapNode(callee.node, t.isMemberExpression);
  if (
    trueMember
      && !trueMember.computed
      && t.isIdentifier(trueMember.property)
  ) {
    const obj = unwrapNode(trueMember.object, t.isIdentifier);
    if (obj) {
      const binding = path.scope.getBindingIdentifier(obj.name);
      if (binding) {
        const registrations = ctx.registrations.namespace.hooks.get(binding);
        if (registrations) {
          let registration: typeof registrations[0];
          for (let i = 0, len = registrations.length; i < len; i += 1) {
            registration = registrations[i];
            if (registration.name === trueMember.property.name) {
              return registration.type;
            }
          }
        }
      }
    }
    return isHookName(ctx, trueMember.property) ? 'custom' : 'none';
  }
  return 'none';
}
