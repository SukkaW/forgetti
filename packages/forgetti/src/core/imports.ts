import type { ImportRegistration } from './presets';

export const RUNTIME_EQUALS: ImportRegistration = {
  name: '$$equals',
  source: 'forgetti/runtime',
  kind: 'named',
};

export const RUNTIME_CACHE: ImportRegistration = {
  name: '$$cache',
  source: 'forgetti/runtime',
  kind: 'named',
};

export const RUNTIME_BRANCH: ImportRegistration = {
  name: '$$branch',
  source: 'forgetti/runtime',
  kind: 'named',
};

export const RUNTIME_MEMO: ImportRegistration = {
  name: '$$memo',
  source: 'forgetti/runtime',
  kind: 'named',
};
