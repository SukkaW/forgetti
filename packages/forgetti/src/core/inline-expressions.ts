/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type * as babel from '@babel/core';
import * as t from '@babel/types';
import type { ComponentNode } from './types';
import { isPathValid } from './checks';

function isInValidExpression(path: babel.NodePath): boolean {
  let current = path.parentPath;
  let prev = path;
  while (current) {
    if (
      isPathValid(current, t.isConditionalExpression)
      && (
        current.get('consequent').node === prev.node
        || current.get('alternate').node === prev.node
      )
    ) {
      return false;
    }
    if (
      isPathValid(current, t.isLogicalExpression)
      && current.get('right').node === prev.node
    ) {
      return false;
    }
    prev = current;
    current = current.parentPath;
  }
  return true;
}

export function inlineExpressions(
  path: babel.NodePath<ComponentNode>,
): void {
  path.traverse({
    Expression(p) {
      if (p.getFunctionParent() === path && isPathValid(p, t.isIdentifier)) {
        const binding = p.scope.getBinding(p.node.name);
        if (binding && binding.referenced && binding.referencePaths.length === 1) {
          switch (binding.kind) {
            case 'const':
            case 'let':
            case 'var': {
              // move the node to the reference
              const ref = binding.referencePaths[0];
              if (
                isInValidExpression(ref)
                && isPathValid(binding.path, t.isVariableDeclarator)
                && binding.path.node.init
                && isPathValid(binding.path.get('id'), t.isIdentifier)
                && binding.path.scope.getBlockParent() === ref.scope.getBlockParent()
              ) {
                ref.replaceWith(binding.path.node.init);
                binding.path.remove();
              }
            }
              break;
            default:
              break;
          }
        }
      }
    },
  });
  path.scope.crawl();
}
