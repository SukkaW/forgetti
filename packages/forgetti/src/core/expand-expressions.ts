/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import * as t from '@babel/types';
import type * as babel from '@babel/core';
import type { ComponentNode, StateContext } from './types';
import { getHookCallType } from './get-hook-call-type';
import { isPathValid } from './checks';

function isStatementValid(path: babel.NodePath): boolean {
  if (path) {
    switch (path.node.type) {
      case 'ForInStatement':
      case 'ForOfStatement':
      case 'ForStatement':
      case 'WhileStatement':
      case 'DoWhileStatement':
        return false;
      default:
        return true;
    }
  }
  return false;
}

function isInValidExpression(path: babel.NodePath): boolean {
  let current = path.parentPath;
  let prev = path;
  while (current) {
    if (
      isPathValid(current, t.isConditionalExpression)
      && (
        current.get('consequent').node === prev.node
        || current.get('alternate').node === prev.node
      )
    ) {
      return false;
    }
    if (
      isPathValid(current, t.isLogicalExpression)
      && current.get('right').node === prev.node
    ) {
      return false;
    }
    prev = current;
    current = current.parentPath;
  }
  return true;
}

export function expandExpressions(
  ctx: StateContext,
  path: babel.NodePath<ComponentNode>,
): void {
  if (path.node.type === 'ArrowFunctionExpression' && path.node.body.type !== 'BlockStatement') {
    path.node.body = t.blockStatement(
      [t.returnStatement(path.node.body)],
    );
  }

  const hoistedVars = new Set();

  path.traverse({
    AssignmentExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (
        parent === path
        && statement
        && isStatementValid(statement)
        && isInValidExpression(p)
        && !(
          isPathValid(p.parentPath, t.isVariableDeclarator)
          && p.parentPath.node.id.type === 'Identifier'
          && hoistedVars.has(p.parentPath.node.id.name)
        )
      ) {
        const id = p.scope.generateUidIdentifier('hoisted');
        hoistedVars.add(id.name);
        statement.insertBefore(
          t.variableDeclaration(
            'let',
            [t.variableDeclarator(id, p.node)],
          ),
        );
        p.replaceWith(id);
      }
    },
    CallExpression(p) {
      const parent = p.getFunctionParent();
      const statement = p.getStatementParent();

      if (
        parent === path
        && statement
        && isStatementValid(statement)
        && !(
          isPathValid(p.parentPath, t.isVariableDeclarator)
          && p.parentPath.node.id.type === 'Identifier'
          && hoistedVars.has(p.parentPath.node.id.name)
        )
      ) {
        const hookType = getHookCallType(ctx, p);
        if (hookType === 'custom' || hookType === 'effect') {
          const id = p.scope.generateUidIdentifier('hoisted');
          hoistedVars.add(id.name);
          statement.insertBefore(
            t.variableDeclaration(
              'let',
              [t.variableDeclarator(id, p.node)],
            ),
          );
          p.replaceWith(id);
        }
      }
    },
  });

  path.scope.crawl();
}
