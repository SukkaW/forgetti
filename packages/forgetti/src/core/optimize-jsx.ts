import type * as babel from '@babel/core';
import * as t from '@babel/types';
import type { ComponentNode, StateContext } from './types';
import getImportIdentifier from './get-import-identifier';
import { RUNTIME_MEMO } from './imports';
import { isNodeShouldBeSkipped } from './checks';

interface JSXReplacement {
  id: t.Identifier;
  value: t.Expression;
}

interface State {
  // Identifier of the value argument
  source: t.Identifier;
  // List of dependency expressions
  expressions: t.Expression[];
  // List of JSX replacements
  jsxs: JSXReplacement[];
  // key
  key?: t.Expression;
}

function extractJSXExpressions(
  path: babel.NodePath<t.JSXElement | t.JSXFragment>,
  state: State,
  top: boolean,
): void {
  // Iterate attributes
  if (path.isJSXElement()) {
    const openingElement = path.get('openingElement');
    let name = openingElement.get('name');
    while (name.isJSXMemberExpression()) {
      name = name.get('object');
    }
    if (name.isJSXIdentifier()) {
      if (/^[A-Z_]/.test(name.node.name)) {
        const id = path.scope.generateUidIdentifier('Component');
        state.jsxs.push({
          id,
          value: t.identifier(name.node.name),
        });
        name.replaceWith(t.jsxIdentifier(id.name));
      }
    }
    const attrs = openingElement.get('attributes');
    for (let i = 0, len = attrs.length; i < len; i++) {
      const attr = attrs[i];

      if (attr.isJSXAttribute()) {
        const key = attr.get('name');
        if (top && key.isJSXIdentifier() && key.node.name === 'key') {
          const value = attr.get('value');
          if (value.isExpression()) {
            state.key = value.node;
          } else if (value.isJSXExpressionContainer()) {
            const expr = value.get('expression');
            if (expr.isExpression()) {
              state.key = expr.node;
              attr.remove();
            }
          }
        } else {
          const value = attr.get('value');
          if (value.isJSXElement() || value.isJSXFragment()) {
            extractJSXExpressions(value, state, false);
          }
          if (value.isJSXExpressionContainer()) {
            const expr = value.get('expression');
            if (expr.isJSXElement() || expr.isJSXFragment()) {
              extractJSXExpressions(expr, state, false);
            } else if (expr.isExpression()) {
              const id = state.expressions.length;
              state.expressions.push(expr.node);
              expr.replaceWith(
                t.memberExpression(
                  state.source,
                  t.numericLiteral(id),
                  true,
                ),
              );
            }
          }
        }
      }
      if (attr.isJSXSpreadAttribute()) {
        const arg = attr.get('argument');
        if (arg.isJSXElement() || arg.isJSXFragment()) {
          extractJSXExpressions(arg, state, false);
        } else {
          const id = state.expressions.length;
          state.expressions.push(arg.node);
          arg.replaceWith(
            t.memberExpression(
              state.source,
              t.numericLiteral(id),
              true,
            ),
          );
        }
      }
    }
  }

  const children = path.get('children');
  for (let i = 0, len = children.length; i < len; i++) {
    const child = children[i];

    if (child.isJSXElement() || child.isJSXFragment()) {
      extractJSXExpressions(child, state, false);
    } else if (child.isJSXExpressionContainer()) {
      const expr = child.get('expression');
      if (expr.isJSXElement() || expr.isJSXFragment()) {
        extractJSXExpressions(expr, state, false);
      } else if (expr.isExpression()) {
        const id = state.expressions.length;
        state.expressions.push(expr.node);
        expr.replaceWith(
          t.memberExpression(
            state.source,
            t.numericLiteral(id),
            true,
          ),
        );
      }
    } else if (child.isJSXSpreadChild()) {
      const arg = child.get('expression');
      if (arg.isJSXElement() || arg.isJSXFragment()) {
        extractJSXExpressions(arg, state, false);
      } else {
        const id = state.expressions.length;
        state.expressions.push(arg.node);
        arg.replaceWith(
          t.memberExpression(
            state.source,
            t.numericLiteral(id),
            true,
          ),
        );
      }
    }
  }
}

function transformJSX(
  ctx: StateContext,
  path: babel.NodePath<t.JSXElement | t.JSXFragment>,
): void {
  if (isNodeShouldBeSkipped(path.node)) {
    return;
  }
  const state: State = {
    source: path.scope.generateUidIdentifier('values'),
    expressions: [],
    jsxs: [],
  };
  extractJSXExpressions(path, state, true);

  const memoComponent = path.scope.generateUidIdentifier('Memo');

  let body: t.Expression | t.BlockStatement;
  if (state.jsxs.length) {
    const declarations: t.VariableDeclarator[] = [];
    for (let i = 0, len = state.jsxs.length; i < len; i++) {
      declarations.push(t.variableDeclarator(
        state.jsxs[i].id,
        state.jsxs[i].value,
      ));
    }
    body = t.blockStatement([
      t.variableDeclaration('const', declarations),
      t.returnStatement(path.node),
    ]);
  } else {
    body = path.node;
  }
  path.scope.getProgramParent().push({
    kind: 'const',
    id: memoComponent,
    init: t.callExpression(
      getImportIdentifier(
        ctx,
        path,
        RUNTIME_MEMO,
      ),
      [
        getImportIdentifier(
          ctx,
          path,
          ctx.preset.runtime.memo,
        ),
        t.arrowFunctionExpression(
          [state.source],
          body,
        ),
      ],
    ),
  });

  const attrs = [];

  if (state.key) {
    attrs.push(
      t.jsxAttribute(
        t.jsxIdentifier('key'),
        t.jsxExpressionContainer(
          state.key,
        ),
      ),
    );
  }

  attrs.push(
    t.jsxAttribute(
      t.jsxIdentifier('v'),
      t.jsxExpressionContainer(
        t.arrayExpression(state.expressions),
      ),
    ),
  );

  path.replaceWith(
    t.addComment(
      t.jsxElement(
        t.jsxOpeningElement(
          t.jsxIdentifier(memoComponent.name),
          attrs,
          true,
        ),
        t.jsxClosingElement(
          t.jsxIdentifier(memoComponent.name),
        ),
        [],
        true,
      ),
      'leading',
      '@forgetti skip',
      false,
    ),
  );
}

export default function optimizeJSX(
  ctx: StateContext,
  path: babel.NodePath<ComponentNode>,
): void {
  path.traverse({
    JSXElement(p) {
      transformJSX(ctx, p);
    },
    JSXFragment(p) {
      transformJSX(ctx, p);
    },
  });
  path.scope.crawl();
}
