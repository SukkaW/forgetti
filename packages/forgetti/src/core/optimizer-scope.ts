/* eslint-disable @typescript-eslint/no-this-alias */
import * as t from '@babel/types';
import type { OptimizedExpression, StateContext } from './types';
import getImportIdentifier from './get-import-identifier';
import { RUNTIME_BRANCH, RUNTIME_CACHE } from './imports';

function mergeVariableDeclaration(statements: t.Statement[]): t.Statement[] {
  let stack: t.VariableDeclarator[] = [];
  const newStatements: t.Statement[] = [];
  let value: t.Statement;
  for (let i = 0, len = statements.length; i < len; i++) {
    value = statements[i];
    if (value.type === 'VariableDeclaration' && value.kind === 'let') {
      stack.push(...value.declarations);
    } else {
      if (stack.length) {
        newStatements.push(t.variableDeclaration('let', stack));
        stack = [];
      }
      newStatements.push(value);
    }
  }
  return newStatements;
}

export default class OptimizerScope {
  memo: t.Identifier | undefined;

  indeces = 0;

  ctx: StateContext;

  path: babel.NodePath;

  parent?: OptimizerScope;

  isInLoop?: boolean;

  statements: t.Statement[] = [];

  constructor(
    ctx: StateContext,
    path: babel.NodePath,
    parent?: OptimizerScope,
    isInLoop?: boolean,
  ) {
    this.ctx = ctx;
    this.path = path;
    this.parent = parent;
    this.isInLoop = isInLoop;
  }

  createHeader(): t.Identifier {
    if (!this.memo) {
      this.memo = this.path.scope.generateUidIdentifier('c');
    }
    return this.memo;
  }

  createIndex(): t.NumericLiteral {
    const current = this.indeces;
    this.indeces += 1;
    return t.numericLiteral(current);
  }

  getMemoDeclaration(): t.VariableDeclaration | undefined {
    if (!this.memo) {
      return undefined;
    }
    // This is for generating branched caching.
    // Parent means that we want to create the cache
    // from the parent (or root)
    if (this.parent) {
      const header = this.parent.createHeader();
      const index = this.parent.createIndex();

      return t.variableDeclaration('let', [
        t.variableDeclarator(
          this.createHeader(),
          t.callExpression(
            getImportIdentifier(
              this.ctx,
              this.path,
              RUNTIME_BRANCH,
            ),
            [header, index, t.numericLiteral(this.indeces)],
          ),
        ),
      ]);
    }
    return t.variableDeclaration('let', [
      t.variableDeclarator(
        this.memo,
        t.callExpression(
          getImportIdentifier(
            this.ctx,
            this.path,
            RUNTIME_CACHE,
          ),
          [
            getImportIdentifier(
              this.ctx,
              this.path,
              this.ctx.preset.memo,
            ),
            t.numericLiteral(this.indeces),
          ],
        ),
      ),
    ]);
  }

  loop: t.Identifier | undefined;

  loopIndex: t.Identifier | undefined;

  createLoopHeader(): t.Identifier {
    if (!this.loop) {
      this.loop = this.path.scope.generateUidIdentifier('l');
    }
    return this.loop;
  }

  createLoopIndex(): t.Identifier {
    if (!this.loopIndex) {
      this.loopIndex = this.path.scope.generateUidIdentifier('id');
    }
    return this.loopIndex;
  }

  getLoopMemoDeclaration(): t.VariableDeclaration | undefined {
    if (!this.parent) {
      return undefined;
    }
    const header = this.parent.createHeader();
    const index = this.parent.createIndex();
    const id = this.createLoopIndex();

    return t.variableDeclaration('let', [
      t.variableDeclarator(
        this.createHeader(),
        t.callExpression(
          getImportIdentifier(
            this.ctx,
            this.path,
            RUNTIME_BRANCH,
          ),
          // Looped branches cannot be statically analyzed
          [header, index, t.numericLiteral(0)],
        ),
      ),
      // This is for tracking the dynamic size
      t.variableDeclarator(id, t.numericLiteral(0)),
    ]);
  }

  getLoopDeclaration(): t.VariableDeclaration {
    const header = this.createHeader();
    const index = this.createLoopIndex();
    const localIndex = this.path.scope.generateUidIdentifier('lid');
    return t.variableDeclaration('let', [
      t.variableDeclarator(localIndex, t.updateExpression('++', index)),
      t.variableDeclarator(
        this.createLoopHeader(),
        t.callExpression(
          getImportIdentifier(
            this.ctx,
            this.path,
            RUNTIME_BRANCH,
          ),
          [header, localIndex, t.numericLiteral(this.indeces)],
        ),
      ),
    ]);
  }

  getStatements(): t.Statement[] {
    const result = [...this.statements];
    const header = this.isInLoop
      ? this.getLoopDeclaration()
      : this.getMemoDeclaration();
    if (header) {
      return mergeVariableDeclaration([
        header,
        ...result,
      ]);
    }
    return mergeVariableDeclaration(result);
  }

  push(...statements: t.Statement[]): void {
    this.statements = this.statements.concat(statements);
  }

  optimizedID = new WeakMap<t.Identifier, OptimizedExpression>();

  setOptimized(key: t.Identifier, value: OptimizedExpression): void {
    this.optimizedID.set(key, value);
  }

  getOptimized(key: t.Identifier): OptimizedExpression | undefined {
    // let current: OptimizerScope | undefined = this;
    // while (current) {
    //   const result = current.optimizedID.get(key);
    //   if (result) {
    //     return result;
    //   }
    //   current = current.parent;
    // }
    return this.optimizedID.get(key);
  }

  deleteOptimized(key: t.Identifier): void {
    // let current: OptimizerScope | undefined = this;
    // while (current) {
    //   if (current.optimizedID.has(key)) {
    //     current.optimizedID.delete(key);
    //     return;
    //   }
    //   current = current.parent;
    // }
    if (this.optimizedID.has(key)) {
      this.optimizedID.delete(key);
    }
  }

  constants = new WeakSet<t.Identifier>();

  addConstant(value: t.Identifier): void {
    this.constants.add(value);
  }

  hasConstant(value: t.Identifier): boolean {
    // let current: OptimizerScope | undefined = this;
    // while (current) {
    //   const result = current.constants.has(value);
    //   if (result) {
    //     return result;
    //   }
    //   current = current.parent;
    // }
    // return false;
    return this.constants.has(value);
  }
}
