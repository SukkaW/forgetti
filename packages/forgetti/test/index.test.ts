/* eslint-disable no-template-curly-in-string */
import * as babel from '@babel/core';
import { describe, it } from 'vitest';
import type { Options } from '../src';
import plugin from '../src';

const options: Options = {
  preset: 'react',
};

async function compile(code: string): Promise<string> {
  const result = await babel.transformAsync(code, {
    plugins: [
      [plugin, options],
    ],
    parserOpts: {
      plugins: [
        'jsx',
      ],
    },
  });

  return result?.code ?? '';
}

describe.concurrent('statements', () => {
  it('should optimize for-of statements', async ({ expect }) => {
    const code = `
  function Example(props) {
    return props.a && props.b && props.c && props.d;
  }
  `;
    expect(await compile(code)).toMatchSnapshot();
  });
});
