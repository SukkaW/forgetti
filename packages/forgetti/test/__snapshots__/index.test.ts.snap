// Vitest Snapshot v1

exports[`statements > should optimize for-of statements 1`] = `
"import { useRef as _useRef } from \\"react\\";
import { $$cache as _$$cache } from \\"forgetti/runtime\\";
import { $$branch as _$$branch } from \\"forgetti/runtime\\";
import { $$equals as _$$equals } from \\"forgetti/runtime\\";
function Example(props) {
  let _cache = _$$cache(_useRef, 5),
    _equals = _$$equals(_cache, 0, props),
    _value4 = _equals ? _cache[0] : _cache[0] = props,
    _value5 = _equals ? _cache[1] : _cache[1] = _value4.a,
    _value3;
  if (_value5) {
    let _cache2 = _$$branch(_cache, 2, 2),
      _equals2 = _$$equals(_cache2, 0, props),
      _value6 = _equals2 ? _cache2[0] : _cache2[0] = props;
    _value3 = _equals2 ? _cache2[1] : _cache2[1] = _value6.b;
  } else _value3 = _value5;
  let _value2;
  if (_value3) {
    let _cache3 = _$$branch(_cache, 3, 2),
      _equals3 = _$$equals(_cache3, 0, props),
      _value8 = _equals3 ? _cache3[0] : _cache3[0] = props;
    _value2 = _equals3 ? _cache3[1] : _cache3[1] = _value8.c;
  } else _value2 = _value3;
  let _value;
  if (_value2) {
    let _cache4 = _$$branch(_cache, 4, 2),
      _equals4 = _$$equals(_cache4, 0, props),
      _value10 = _equals4 ? _cache4[0] : _cache4[0] = props;
    _value = _equals4 ? _cache4[1] : _cache4[1] = _value10.d;
  } else _value = _value2;
  return _value;
}"
`;
